lib/benchmark/benchmark.go

```diff
@@ -149,157 +157,133 @@ func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Resu
 	tc.Stdout = ioutil.Discard
 	tc.Stderr = ioutil.Discard
 	tc.Stdin = &bytes.Buffer{}
-	ctx, cancelWorkers := context.WithCancel(ctx)
-	defer cancelWorkers()
+	var delay time.Duration = 0
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	requestC := make(chan benchMeasure)
-	responseC := make(chan benchMeasure, c.Threads)
+	requestsC := make(chan benchMeasure)
+	resultC := make(chan benchMeasure)
 
-	for i := 0; i < c.Threads; i++ {
-		thread := &benchmarkThread{
-			id:          i,
-			ctx:         ctx,
-			client:      tc,
-			command:     c.Command,
-			interactive: c.Interactive,
-			receiveC:    requestC,
-			sendC:       responseC,
+	go func() {
+		interval := time.Duration(1 / float64(c.Rate) * float64(time.Second))
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		start := time.Now()
+		for {
+			select {
+			case <-ticker.C:
+				// ticker makes its first tick after the given duration, not immediately
+				// this sets the send measure ResponseStart time accurately
+				delay = delay + interval
+				t := start.Add(delay)
```

Comment:
If this producer loop blocks for longer than the tick rate for any reason (e.g. the `requestsC <- measure` send) `time.Now()` will no longer be accurate -- resulting in coordinated omission.

Thus we prefer using math to calculate start time as currently equivalent, and more future proof. E.g. if we add back in `--threads` or some logic to handle throttling/back-pressure based on clientside resource (fh?) exhaustion, the math based approach remains correct.

Based on your comment, I worry that this subtlety may be lost on future maintainers, as response vs service time is not as explicitly called out anymore.

Reply:
ahh, in that case you can use the value from `ticker.C`, which emits `time.Time` value for when the tick was emitted

Reply:
`ticker.C` will still result in coordinated omission.  Per the [docs](https://golang.org/pkg/time/#NewTicker):

"[Ticker] adjusts the intervals or drops ticks to make up for slow receivers."

For a brief illustration:

```
$ cat tickertest.go   
package main

import (
        "context"
        "log"
        "time"
)

func main() {
        ctx, cancel := context.WithCancel(context.TODO())
        defer cancel()
        go func() {
                time.Sleep(5 * time.Second)
                cancel()
        }()
        go func() {
                ticker := time.NewTicker(250 * time.Millisecond)
                for {
                        select {
                        case t := <-ticker.C:
                                log.Println("tick at ", t)
                                time.Sleep(1 * time.Second)
                        case <-ctx.Done():
                                return
                        }
                }
        }()
        <-ctx.Done()
}
$ go run tickertest.go
2020/11/30 12:21:41 tick at  2020-11-30 12:21:41.879970194 -0800 PST m=+0.250283039
2020/11/30 12:21:42 tick at  2020-11-30 12:21:42.129849762 -0800 PST m=+0.500162518
2020/11/30 12:21:43 tick at  2020-11-30 12:21:43.130051902 -0800 PST m=+1.500364734
2020/11/30 12:21:44 tick at  2020-11-30 12:21:44.130045267 -0800 PST m=+2.500358106
2020/11/30 12:21:45 tick at  2020-11-30 12:21:45.12999193 -0800 PST m=+3.500304768
$
```

Reply:
We could do something like this and there wouldn't be coordinated omission, but it's essentuially the same. It sounds like we really want to use ticker.C though 

```
	go func() {
		interval := time.Duration(1 / float64(c.Rate) * float64(time.Second))
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				// ticker makes its first tick after the given duration, not immediately
				// this sets the send measure ResponseStart time accurately
				delay = delay + interval
				tickerTime := <-ticker.C
				t := tickerTime.Add(-delay)
				measure := benchMeasure{
					ResponseStart: t,
					command:       c.Command,
					client:        tc,
					interactive:   c.Interactive,
				}
				go work(ctx, measure, resultC)
			case <-ctx.Done():
				close(requestsC)
				return
			}
		}
	}()
```

Reply:
Hmm... if you miss ticker events, aren't you then mis-representing your RPS?

For example: your rate is 10 rps, and you miss 1 ticker event (due to 1 slow iteration). Your actual RPS is 9 (for that second) and the calculated start time for all measurements after the missed event are increased by 100ms.
If you miss 1 event every second, you start accumulating this calculated start time gap by 100ms every 1s.
After 10s, even if everything returns to normal with no server slowness or missed events, all your measurements are increased by 1s because of 10 missed ticks.

I think there are 2 possible approaches:
1. detect missed ticks (by checking elapsed time since last tick) and mark those measurements as failed
2. make the ticker loop extremely fast, independent of server response time so that ticks are never dropped

The 2nd option is what you already have: each iteration creates a measurement object and kicks of a goroutine with it. This should consistently run fast and not cause ticks to be missed.

Please let me know if I'm mis-understanding coordinated omission though :)

Reply:
https://play.golang.org/p/fLMrlvh222b illustrates what I mean by "accumulating start time gap"

Reply:
So coordinated omission just means you are not correctly recording response time. There aren't any ticks that are being missed, I am just simply starting them when they are mathematically supposed to be started. 

In your example, you are adding delay. I am subtracting it when using ticker in the example above. But your right, what I already have works cause it is an open system and does not depend on anything else

Reply:
You're right: if the current implementation falls behind (it shouldn't :crossed_fingers:) it would never catch up. At best it would return back to the expected RPS with a fixed delay. However, in the event this happens, is better to measure & report on that delay -- that will show up as a red flag to investigate.

> 2. make the ticker loop extremely fast, independent of server response time so that ticks are never dropped

I agree that we prefer the current implementation.  Downsides:

 - OS contention (I think @quinqu was hitting filehandle/socket limits when pushing `tsh bench` to demonstrate coordinated omission)
 - goroutines become a poor unbounded & unordered queue, and will pile up if the response rate < request rate, and requests aren't failing for other reasons (above OS limits, teleport backpressure).  We're creating an expected leak.  Not that testing RPS >> limits is a particular useful perf test to run... but it is expected during exploratory/redlines testing.

Back to the original question that started this thread: Should we record the scheduled start of the request or the observed start of the request?  I personally prefer both.  The scheduled start time should be used for response time latency calculation.  The observed time can be used to measure service time (which should always be equal to response time in a system without backpressure).

If we only record one, scheduled start & response time is better, as service time ignores throttling.  At best `time.Now()` is equivalent to `start + delay`. At its worst using `time.Now()` obscures perf issues.

Reply:
I see what you mean by your example though

Reply:
> https://play.golang.org/p/fLMrlvh222b illustrates what I mean by "accumulating start time gap"

This nicely illustrates the difference between service time & response time.  In your exampe the last request has 5ms service time and 105ms response time.  Coordinated omission is the claim that service time is response time.

Reply:
P.S.  This is a lot of back & forth for a PR comment.  If we still need to hash out a consensus, we might do have a video call to get on the same page.

Reply:
I'm okay with any solution you pick.
Just wanted to point out that manually calculating tick times seemed redundant and bug-prone.

As for unbounded goroutine leak - there should be some hard timeout on all commands (e.g. 30s or something longer). But even without it, it's ok since this is a client-side tool and not a backend system stealing resources from other critical software.

Reply:
To me, it makes sense to keep just response time because that is the only thing we _should_ care about. Maybe a feature (down the line) could be added to show both if specified, but only response by default

Reply:
I think if we were to keep both and a user didn't know the difference, service time could be valued over response time since it "looks" better, that is just my thought process.